cpp/src/arrow/builder.h

```diff
@@ -139,6 +144,38 @@ class ARROW_EXPORT ArrayBuilder {
     ++length_;
   }
 
+  template <typename IterType>
+  void UnsafeAppendToBitmap(const IterType& begin, const IterType& end) {
+    int64_t byte_offset = length_ / 8;
+    int64_t bit_offset = length_ % 8;
+    uint8_t bitset = null_bitmap_data_[byte_offset];
+
+    for (auto iter = begin; iter != end; ++iter) {
```

Comment:
As an aside, would you expect the optimized code to be significantly different here vs. the original `const uint8_t* valid_bytes` version?

Reply:
With STL iterators and raw pointers it should be exactly the same. With more complicated iterators, it's hard to tell, though the `LazyRange` ones the same performance as loops, at least on GCC.